vault/resource_managed_keys.go

```diff
@@ -0,0 +1,661 @@
+	if e != nil {
+		return e
+	}
+
+	var keyType string
+	switch providerType {
+	case consts.FieldAWS:
+		keyType = kmsTypeAWS
+
+	case consts.FieldPKCS:
+		keyType = kmsTypePKCS
+
+	case consts.FieldAzure:
+		keyType = kmsTypeAzure
+
+	default:
+		return fmt.Errorf("received unexpected provider type %s", providerType)
+	}
+
+	p := fmt.Sprintf("%s/%s", "sys/managed-keys", keyType)
+	log.Printf("[DEBUG] Listing data from Vault at %s", p)
+	resp, err := client.Logical().List(p)
+	if err != nil {
+		return err
+	}
+
+	if resp == nil {
+		return nil
+	}
+
+	handlers := map[string]func() schemaMap{
+		kmsTypeAWS:   managedKeysAWSConfigSchema,
+		kmsTypePKCS:  managedKeysPKCSConfigSchema,
+		kmsTypeAzure: managedKeysAzureConfigSchema,
+	}
+
+	var data []interface{}
+	if v, ok := resp.Data["keys"]; ok {
+		for _, name := range v.([]interface{}) {
+			m := make(map[string]interface{})
+			path := getManagedKeysPath(keyType, name.(string))
+			log.Printf("[DEBUG] Reading from Vault at %s", path)
+			resp, err := client.Logical().Read(path)
+			if err != nil {
+				return err
+			}
+
+			if resp == nil {
+				continue
+			}
+
+			for k := range handlers[keyType]() {
+				// Map TF schema fields to Vault API
+				vaultKey := k
+				if v, ok := sm[k]; ok {
+					vaultKey = v
+				}
+
+				if v, ok := resp.Data[vaultKey]; ok {
+					m[k] = v
+				}
+			}
+
+			// get these from TF config since they are
+			// returned as "redacted" from Vault
+			m = getRedactedFields(d, providerType, name.(string), redactedFields, m)
+
+			data = append(data, m)
+		}
+	}
+
+	s := schema.NewSet(hashManagedKeys, data)
+
+	if err := d.Set(providerType, s); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func readAWSManagedKeys(d *schema.ResourceData, meta interface{}) error {
+	redacted := []string{"access_key", "secret_key"}
+	sm := map[string]string{
+		consts.FieldUUID: "UUID",
+	}
+	if err := readAndSetManagedKeys(d, meta, consts.FieldAWS, sm, redacted); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func readAzureManagedKeys(d *schema.ResourceData, meta interface{}) error {
+	var redacted []string
+	sm := map[string]string{
+		consts.FieldUUID: "UUID",
+	}
+	if err := readAndSetManagedKeys(d, meta, consts.FieldAzure, sm, redacted); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func readPKCSManagedKeys(d *schema.ResourceData, meta interface{}) error {
+	redacted := []string{"pin"}
+	sm := map[string]string{
+		consts.FieldUUID: "UUID",
+	}
+	if err := readAndSetManagedKeys(d, meta, consts.FieldPKCS, sm, redacted); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func readManagedKeys(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	diags := diag.Diagnostics{}
+
+	err := readAWSManagedKeys(d, meta)
+	if err != nil {
+		diags = append(diags, diag.Diagnostic{
+			Severity: diag.Error,
+			Summary:  fmt.Sprintf("Failed to read AWS Managed Keys, err=%s", err),
+		})
+	}
+
+	err = readPKCSManagedKeys(d, meta)
```

Comment:
nit: we can avoid the shadowed `err`  if we change these to be of the form `if err := .... ; err != nil {...}`